File: Problems/Algorithms/239. Sliding Window Maximum/max_sliding_window.py
from collections import deque


class Solution:
    def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:
        queue = deque()

        for i in range(k):
            while len(queue) > 0:
                if nums[i] > nums[queue[-1]]:
                    queue.pop()
                else:
                    break
            queue.append(i)

        results = []
        for i in range(k, len(nums)):
            results.append(nums[queue[0]])

            if queue[0] < i - k + 1:
                queue.popleft()

            while len(queue) > 0:
                if nums[i] > nums[queue[-1]]:
                    queue.pop()
                else:
                    break
            queue.append(i)
        results.append(nums[queue[0]])

        return results

File: Problems/Algorithms/239. Sliding Window Maximum/MaxSlidingWindow.kt
class Solution {
    fun maxSlidingWindow(nums: IntArray, k: Int): IntArray {
        val queue: MutableList<Int> = mutableListOf()
        val results = IntArray(nums.size-k+1) { 0 }
        
        for (i in 0..k-1) {
            while (!queue.isEmpty()) {
                if (nums[i] > nums[queue.get(queue.size-1)]) {
                    queue.removeAt(queue.size-1)
                } else {
                    break
                }
            }
            queue.add(i)
        }
        
        var j = 0
        for (i in k..nums.size-1) {
            results[j] = nums[queue.get(0)]
            j++
            
            if (queue.get(0) < i-k+1) {
                queue.removeAt(0)
            }
            
            while (!queue.isEmpty()) {
                if (nums[i] > nums[queue.get(queue.size-1)]) {
                    queue.removeAt(queue.size-1)
                } else {
                    break
                }
            }
            
            queue.add(i)
        }
        results[j] = nums[queue.get(0)]
        
        return results
    }
}

